src/core/event/poll_group.cpp

```diff
@@ -154,3 +161,19 @@ void poll_group::del_ring(ring *rng)
         m_rings.erase(iter);
     }
 }
+
+void poll_group::add_socket(sockinfo_tcp *si)
+{
+    m_sockets_list.push_back(si);
+}
+
+void poll_group::close_socket(sockinfo_tcp *si, bool force /*=false*/)
+{
+    m_sockets_list.erase(si);
+
+    bool closed = si->prepare_to_close(force);
+    if (closed) {
+        si->clean_socket_obj();
+    }
+    // TODO If not closed, the socket will be destroyed after the last completion notification.
```

Comment:
Should it be something that needs still be handled? Will the socket still live in timers pending to remove?

Reply:
This is TODO for current PR. We need to keep the socket object until all the TX completions are handled. This is different then pending to remove, because in current case, we are ready to destroy the socket object.

Reply:
To reduce the chance of this scenario, polling group method does a polling iteration to collect remaining TX completions. Usually a socket is closed in a state where there are no pending data. But even if a socket is closed in the middle of TCP stream, we force TCP reset, so the only problematic scenario is when there WQEs in SQ that are not completed until to polling group poll() inside the close method. This scenario will be addressed in future version.